primitives/src/sybil.rs

```diff
@@ -1,34 +1,100 @@
-use crate::communities::CommunityIdentifier;
 use crate::scheduler::CeremonyIndexType;
 
-pub type ProofOfPersonhoodRequest<Signature, AccountId> =
-    Vec<(CommunityIdentifier, ProofOfAttendance<Signature, AccountId>)>;
+#[derive(Encode, Decode, Clone, PartialEq, Eq, Default, RuntimeDebug)]
+pub struct IssueProofOfPersonhoodConfidenceCall<Signature, AccountId> {
+    call_index: [u8; 2],
+    requester: AccountId,
+    request: Vec<ProofOfAttendance<Signature, AccountId>>,
+    requested_response: u8,
+    sender_pallet_index: u8,
+}
+
+impl<Signature, AccountId> IssueProofOfPersonhoodConfidenceCall<Signature, AccountId> {
+    pub fn new(
+        sybil_proof_issuer_index: u8,
+        requester: AccountId,
+        request: Vec<ProofOfAttendance<Signature, AccountId>>,
+        requested_response: RequestedSybilResponse,
+        sender_pallet_index: u8,
+    ) -> Self {
+        Self {
+            call_index: [sybil_proof_issuer_index, 0], // is the first call in proof-issuer pallet
+            requester,
+            request,
+            requested_response: requested_response as u8,
+            sender_pallet_index,
+        }
+    }
+}
+
+/// This allows to generically call the sybil-proof-issuer, whose response calls the method with the
+/// index defined in the `RequestedSybilResponse`
+#[derive(Encode, Decode, Clone, PartialEq, Eq, RuntimeDebug)]
+pub enum RequestedSybilResponse {
+    Faucet = 1,
+}
+
+impl Default for RequestedSybilResponse {
+    fn default() -> RequestedSybilResponse {
+        RequestedSybilResponse::Faucet
+    }
+}
 
-pub type IssueProofOfPersonhoodConfidenceCall<Signature, AccountId> = (
-    [u8; 2],
-    u32,
-    u8,
-    ProofOfPersonhoodRequest<Signature, AccountId>,
-);
+#[derive(Encode, Decode, Clone, PartialEq, Eq, Default, RuntimeDebug)]
+pub struct SybilResponseCall<AccountId> {
+    call_index: [u8; 2],
+    account: AccountId,
+    confidence: ProofOfPersonhoodConfidence,
+}
 
-pub type SetProofOfPersonHoodCall<AccountId> = ([u8; 2], AccountId, ProofOfPersonhoodConfidence);
+impl<AccountId> SybilResponseCall<AccountId> {
+    pub fn new(
+        sybil_gate_index: u8,
+        requested_sybil_response_call_index: u8,
+        account: AccountId,
+        confidence: ProofOfPersonhoodConfidence,
+    ) -> Self {
+        Self {
+            call_index: [sybil_gate_index, requested_sybil_response_call_index],
+            account,
+            confidence,
+        }
+    }
+}
 
-#[derive(Encode, Decode, Copy, Clone, PartialEq, Eq, Default, RuntimeDebug)]
+#[derive(Encode, Decode, Clone, PartialEq, Eq, Default, RuntimeDebug)]
 pub struct ProofOfPersonhoodConfidence {
     attested: CeremonyIndexType,
     last_n_ceremonies: CeremonyIndexType,
+    proofs: Vec<H256>,
 }
 
 impl ProofOfPersonhoodConfidence {
-    pub fn new(attested: CeremonyIndexType, last_n_ceremonies: CeremonyIndexType) -> Self {
+    pub fn new(
+        attested: CeremonyIndexType,
+        last_n_ceremonies: CeremonyIndexType,
```

Comment:
the return value for the facuet chain can be rounded to a u32 with full scale eq 1.0, no need to depend on fixed for this

future: communities will have a certain confidence rating too, so I suggest that we already introduce \beta as the confidence of the community as such, which auto-initializes to some very small positive number when the community is registered (zero is bad for testing as the confidence will always be zero)
so, the confidence formula becomes

confidence = sum(\beta * exp(T*\alpha)) | for all attended ceremonies with a valid proof

Reply:
I must say, I like the idea of the exponential function.

However, your observation is not correct (the naming is arguably bad). But I do allow gaps. `last_n_ceremonies` stands for the earliest ceremony, a proof was included, and `attested` is the amount of those `last_n_ceremonies`, which a proof was supplied for.

Hence, the idea was that the recipient of the proof has two pieces of information:
* the earliest ceremony that was attended (current_ceremony_index -`n`)
* how many of the last `n` ceremonies were attended.

I thought that this is even more flexible than an exponentially decaying weight of attended ceremonies.

Reply:
Ah, I missed that. Indeed, what you did is sufficient for now. I suggest to open a new issue for the exponential stuff because it will later allow the weighting of community web of trust as well. I'll just copy-paste the idea into an issue.

Reply:
https://github.com/encointer/pallets/issues/24